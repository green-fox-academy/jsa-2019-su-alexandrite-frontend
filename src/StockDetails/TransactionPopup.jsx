import React, { useState } from 'react';
import {
  Text,
  View,
  TouchableHighlight,
  TextInput,
} from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import Popup from '../common/Popup';
import styles from '../Investments/styles';
import TransactionSelector from './TransactionSelector';
import { purchaseStock } from '../redux/stock/actionCreator';

const TransactionPopup = () => {
  const [popupVisible, setModalVisible] = useState(false);
  const [shares, setShares] = useState('');
  const [transactionBehavior, setTransactionBehavior] = useState('sell');
  const { accessToken } = useSelector((state) => state.user);
  const stockName = 'MSFT';
  const dispatch = useDispatch();

  const onCloseAddModal = () => {
    setModalVisible(false);
    setShares('');
  };

  const onConfirmAddModal = () => {
    const share = parseInt(shares, 10);
    if (shares) dispatch(purchaseStock(stockName, share, transactionBehavior, accessToken));
    setModalVisible(false);
    setShares('');
  };

  return (
    <View>
      <Popup
        visible={popupVisible}
        onCancel={onCloseAddModal}
        onConfirm={onConfirmAddModal}
        title={`Purchase ${stockName}`}
        confirmButtonText="OK"
        confirmDisabled={!shares}
      >
        <TextInput style={styles.sharesField} placeholder="shares" onChangeText={(text) => setShares(text)} />
        <TransactionSelector selected={transactionBehavior} onSelect={setTransactionBehavior} />
      </Popup>
      <TouchableHighlight
        onPress={() => setModalVisible(true)}
      >
        <Text>Trade</Text>
      </TouchableHighlight>
    </View>
  );
};

export default TransactionPopup;
