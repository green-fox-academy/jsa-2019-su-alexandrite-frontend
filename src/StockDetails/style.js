import { StyleSheet } from 'react-native';
import { width } from '../common/styles';

export default StyleSheet.create({
  detailsContentContainer: {
    borderRadius: 6,
    paddingHorizontal: 8,
    paddingVertical: 15,
    backgroundColor: '#f2f3fb',
    flexDirection: 'row',
    width: width - 60,
    justifyContent: 'center',
    minHeight: 200,
  },
  detailsContentRow: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  detailsContentKey: {
    fontSize: 11,
    lineHeight: 30,
    letterSpacing: 0.21,
    color: '#7e92a3',
  },
  detailsContentVal: {
    fontSize: 11,
    lineHeight: 30,
    letterSpacing: 0.21,
    color: '#566ed3',
    fontWeight: 'bold',
    textAlign: 'right',
  },
  perfChartContainer: {
    width: width - 60,
    minHeight: 200,
    justifyContent: 'center',
  },
  perfChartTimeRangeSelector: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    backgroundColor: '#f2f3fb',
    marginTop: 10,
    borderRadius: 20,
    padding: 5,
  },
  perfChartButton: {
    paddingHorizontal: 20,
    paddingVertical: 5,
    borderRadius: 20,
  },
  perfChartButtonActive: {
    backgroundColor: '#566ed3',
    elevation: 2,
    shadowColor: '#566ed3',
    shadowRadius: 2,
    shadowOpacity: 0.8,
    shadowOffset: {
      height: 2,
    },
  },
  perfChartButtonText: {
    color: '#666',
    fontWeight: 'bold',
  },
  perfChartButtonTextActive: {
    color: '#fff',
  },
  sharesField: {
    padding: 8,
    borderBottomWidth: 1,
    borderBottomColor: '#eee',
  },
  TransactionSelector: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    backgroundColor: '#f2f3fb',
    marginTop: 10,
    borderRadius: 20,
    padding: 5,
  },
  TransactionButton: {
    paddingHorizontal: 50,
    paddingVertical: 5,
    borderRadius: 10,
  },
  TransactionButtonActive: {
    backgroundColor: '#566ed3',
    elevation: 0,
    shadowColor: '#566ed3',
    shadowRadius: 2,
    shadowOpacity: 0.8,
    shadowOffset: {
      height: 2,
    },
  },
  TransactionButtonText: {
    color: '#666',
    fontWeight: 'bold',
  },
  TransactionButtonTextActive: {
    color: '#fff',
  },
});
