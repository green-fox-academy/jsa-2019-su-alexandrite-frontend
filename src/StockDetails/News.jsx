import React, { useEffect } from 'react';
import {
  View,
  ActivityIndicator,
} from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import { fetchStockNews } from '../redux/stock/actionCreator';
import style from './style';
import Card from '../common/Card';
import ErrorMessage from '../common/ErrorMessage';
import NewsMappedList from '../common/NewsMappedList';

const News = ({ symbol }) => {
  const dispatch = useDispatch();
  const {
    stockNewsData: news,
    loadingStockNews: loading,
    stockNewsError: error,
  } = useSelector((state) => state.stock);

  useEffect(() => {
    dispatch(fetchStockNews(symbol));
  }, []);

  return (
    <Card title="News">
      {!error ? (
        <View style={style.newsContainer}>
          {loading && <ActivityIndicator size="large" />}
          {!loading && news && (
            <NewsMappedList news={news} />
          )}
        </View>
      ) : <ErrorMessage message={error.message} />}
    </Card>
  );
};

News.propTypes = {
  symbol: PropTypes.string.isRequired,
};

export default News;
