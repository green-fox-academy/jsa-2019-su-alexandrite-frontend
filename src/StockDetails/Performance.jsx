import React, { useEffect, useState } from 'react';
import {
  View,
  ActivityIndicator,
} from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import { fetchHistoryChartData, resetStockInfo } from '../redux/stock/actionCreator';
import style from './style';
import Card from '../common/Card';
import ErrorMessage from '../common/ErrorMessage';
import PerformanceChart from './charts/PerformanceChart';
import TimeRangeSelector from './charts/TimeRangeSelector';

const Performance = ({ symbol }) => {
  const dispatch = useDispatch();
  const [timeRange, setTimeRange] = useState('1y');
  const {
    historicalData: data,
    loadingHistoricalChartData: loading,
    historicalDataError: error,
  } = useSelector((state) => state.stock);

  useEffect(() => {
    dispatch(fetchHistoryChartData(symbol, timeRange));
  }, [timeRange]);

  useEffect(
    () => () => dispatch(resetStockInfo()),
    [],
  );

  return (
    <Card title="Performance">
      {!error ? (
        <View style={style.perfChartContainer}>
          {loading && <ActivityIndicator size="large" />}
          {!loading && !error && (
            <PerformanceChart
              data={data}
              range={timeRange}
            />
          )}
        </View>
      ) : <ErrorMessage message={error.message} />}
      <TimeRangeSelector selected={timeRange} onSelect={setTimeRange} />
    </Card>
  );
};

Performance.propTypes = {
  symbol: PropTypes.string.isRequired,
};

export default Performance;
