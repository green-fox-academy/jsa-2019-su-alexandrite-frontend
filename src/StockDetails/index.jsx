import React, { useEffect, useState } from 'react';
import {
  View,
  Image,
  ScrollView,
  TouchableHighlight,
} from 'react-native';
import { useNavigationParam, useNavigation } from 'react-navigation-hooks';
import commonStyles from '../common/styles';
import Stats from './Stats';
import addToWatchlistIcon from '../../assets/icons/header/add-to-wl.png';
import WatchlistPickerPopup from './WatchlistPickerPopup';
import Performance from './Performance';
import NewsCard from '../common/News/NewsCard';
import FAB from '../common/FAB';
import Row from '../common/Row';

const StockDetails = () => {
  const symbol = useNavigationParam('symbol');
  const navigation = useNavigation();
  const [A2WPopupVisible, setA2WPopupVisible] = useState(false);
  const showAddToWatchlistPopup = () => setA2WPopupVisible(true);

  useEffect(() => {
    navigation.setParams({
      showAddToWatchlistPopup,
    });
  }, []);

  return (
    <View style={{ flex: 1, flexDirection: 'row' }}>
      <ScrollView contentContainerStyle={[commonStyles.container, { paddingBottom: 100 }]}>
        <Performance symbol={symbol} />
        <Row>
          <Stats symbol={symbol} />
        </Row>
        <Row flex={1}>
          <NewsCard query={symbol} style={{ flex: 1 }} />
        </Row>
        <WatchlistPickerPopup
          symbol={symbol}
          visible={A2WPopupVisible}
          onClose={() => setA2WPopupVisible(false)}
        />
      </ScrollView>
      <FAB
        onPress={() => { }}
        iconName="exchange-alt"
      />
    </View>
  );
};

StockDetails.navigationOptions = ({ navigation }) => ({
  title: navigation.getParam('symbol'),
  headerRight: () => {
    const fn = navigation.getParam('showAddToWatchlistPopup');
    return (
      <TouchableHighlight
        onPress={fn}
        underlayColor="#0000"
        activeOpacity={0.5}
        style={commonStyles.headerIcon}
      >
        <Image source={addToWatchlistIcon} />
      </TouchableHighlight>
    );
  },
});

export default StockDetails;
