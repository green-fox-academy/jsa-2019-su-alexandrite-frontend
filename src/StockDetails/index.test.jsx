import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import { useDispatch } from 'react-redux';
import { useNavigation } from 'react-navigation-hooks';
import StockDetails from '.';

jest.mock('react-redux');
jest.mock('react-navigation-hooks');
jest.mock('./Performance', () => '<Performance />');
jest.mock('./Stats', () => '<Stats />');
jest.mock('./WatchlistPickerPopup', () => '<WatchlistPickerPopup />');

describe('<StockDetails />', () => {
  beforeEach(() => {
    useDispatch.mockReturnValueOnce(() => { });
    useNavigation.mockReturnValue({
      push: jest.fn(),
      setParams: jest.fn(),
    });
  });

  afterAll(() => {
    jest.resetModules();
  });

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    const wrapper = shallow(<StockDetails />);
    expect(wrapper).not.toBeNull();
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const tree = renderer.create(
      <StockDetails />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should pass show modal function on load', () => {
    const mockSetParams = jest.fn();
    useNavigation.mockReturnValueOnce({
      setParams: mockSetParams,
    });
    const tree = renderer.create(
      <StockDetails />,
    );
    tree.update(<StockDetails />);
    expect(mockSetParams).toHaveBeenCalled();
  });

  // it('Should pass close function to modal', () => {
  //   const mockSetState = jest.fn();
  //   const spy = jest.spyOn(React, 'useState');
  //   // jest.mock('react', () => ({
  //   //   ...React,
  //   //   useState: jest.fn().mockReturnValue([false, mockSetState]),
  //   // }));
  //   // React.mockReturnValueOnce({

  //   // });
  //   const wrapper = shallow(
  //     <StockDetails />,
  //   );
  //   expect(spy).toHaveBeenCalled();
  //   wrapper.find('WatchlistPickerPopup').props().onClose();
  //   expect(mockSetState).toHaveBeenCalledWith(false);
  it('Should pass close function to modal', () => {
    const mockClose = jest.fn();
    const mockGetParams = jest.fn().mockReturnValue(mockClose);
    useNavigation.mockReturnValueOnce({
      getParam: mockGetParams,
    });
    const headerRightButton = StockDetails
      .navigationOptions({ navigation: useNavigation() })
      .headerRight();
    headerRightButton.props.onPress();
    expect(mockClose).toHaveBeenCalled();
  });
});
