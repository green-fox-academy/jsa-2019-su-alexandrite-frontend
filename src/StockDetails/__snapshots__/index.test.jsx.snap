// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StockDetails /> Should render and match the snapshot 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "alignItems": "center",
      "padding": 15,
    }
  }
  style={
    Object {
      "backgroundColor": "#f4f6f8",
      "flex": 1,
    }
  }
>
  <View>
    <<Performance /> />
    <<Stats /> />
    <<NewsCard /> />
    <<WatchlistPickerPopup />
      onClose={[Function]}
      visible={false}
    />
  </View>
</RCTScrollView>
`;
